[ntuple,daos] Adapt I/O backend to DAOSv2 API

glmiotto opened the pull request:
# This Pull request:
Makes changes to ROOT v7 in order to comply with the API for DAOS 2.0.x, affecting the libraries `RPageStorageDaos`, `RDaos` and `libdaos_mock`, as well as the `ntuple_storage_daos` test suite.

## Changes or fixes:
- As of DAOS 2.0, pool service replica ranks are not longer specified in DAOS URIs.  Thus, `daos_pool_connect` now forgoes the `svc_rank_list` argument.
  - Container addresses follow the simplified pattern `daos://($pool)/($container)`. 
  - `RPageStorageDaos` consequently drops parsing enforcement of the previous pattern.
- DAOS 2.0 introduced support for creating pools and containers from a label that is then hashed internally as a UUID. Pool and containers are now identifiable by user-provided labels of up to 127 characters (alphanumeric, colon, period, hyphen or underscore) or by 36-character UUIDs. Labels that match a UUID pattern are explicitly forbidden. 
  - `RDaos` now only supports container creation from user-supplied labels (via `daos_cont_create_with_label`), due to deprecated usage of `daos_cont_create` with user-supplied UUIDs. However, existing containers may still be opened by their UUIDs. 
  - `RPool` and `RContainer` maintain a copy of valid labels (when supplied) and UUIDs (after successfully opening pools and containers respectively)
- Generation of object IDs in `RDaos`: new redundancy, sharding and object type flags; `daos_obj_generate_oid` replaces `daos_obj_generate_id`. 
- The `libdaos_mock` library is updated in lockstep.
- RNTuple backend testing is updated to follow the DAOS 2.0 URI pattern, using labels for default test pool and containers.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #

jblomer commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -238,13 +243,80 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags,
+   if (len == 0 || len > DAOS_PROP_LABEL_MAX_LEN)
+      return false;
+
+   /** Verify that it contains only alphanumeric characters or :.-_ */
```
```suggestion
   // Verify that it contains only alphanumeric characters or :.-_
```

jblomer commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -238,13 +243,80 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags,
+      return false;
+   }
+
+   /** Check to see if it could be a valid UUID */
```
```suggestion
   // Check to see if it could be a valid UUID
```

jblomer commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -238,13 +243,80 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags,
+      return false;
+   }
+
+   /** Check to see if it could be a valid UUID */
```
Perhaps add a comment why we explicitly check for (and forbid) UUIDs

glmiotto replied:
This function is taken wholesale from a more recent version of Daos. The check explicitly forbids UUIDs while validating labels because the API so disambiguates IDs for pools or containers - e.g. `daos_cont_create` takes `uuid_t` and `daos_cont_create_with_label` takes `const char*` and generates a UUID internally. I presume the intention is to prevent the user from creating a container whose label matches the UUID of an existing container. 

Perhaps it is excessive to forbid UUID-like labels in our mock library, since it is now entirely label-based and does not hash them into `uuid_t` internally. 

For our `RDaos` backend layer, it would likely be sensible to support both forms of ID rather than switch to labels exclusively, like in this PR. In that case, I could switch back to representing pool and container IDs with `uuid_t` as long as the correct API functions are called at initialization and the hash is retrieved.

jblomer reviewed (commented):
Very nice! A few nits, in particular whole line comments in the code should use the `//` style (which makes it easier to comment blocks including comments with `/* */`).

I leave the final approval up to @jalopezg-r00t

glmiotto commented:
> Very nice! A few nits, in particular whole line comments in the code should use the `//` style (which makes it easier to comment blocks including comments with `/* */`).
> 
> I leave the final approval up to @jalopezg-r00t

Thanks for the feedback and suggestions, @jblomer! Block comments will be used more judiciously from now on :)

glmiotto commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -238,13 +243,80 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags,
+   int i;
+   bool maybe_uuid = false;
+
+   /** Label cannot be NULL */
```
```suggestion
```

glmiotto commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -238,13 +243,80 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags,
+   if (label == NULL)
+      return false;
+
+   /** Check the length */
```
```suggestion
```

glmiotto commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -238,13 +243,80 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags,
+      bool is_uuid = true;
+      const char *p;
+
+      /** Implement the check directly to avoid uuid_parse() overhead */
```
```suggestion
      // Implement the check directly to avoid uuid_parse() overhead
```

glmiotto commented on tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:
```diff
@@ -392,34 +379,49 @@ int daos_obj_open(daos_handle_t coh, daos_obj_id_t oid, unsigned int mode, daos_
+      return -DER_INVAL;
+
+   oid->hi &= (1ULL << OID_FMT_INTR_BITS) - 1;
+   /**
+    ** | Upper bits contain
+    ** | OID_FMT_TYPE_BITS (object type)
+    ** | OID_FMT_META_BITS (object class metadata)
+    ** | OID_FMT_CLASS_BITS (object class)
+    ** | 96-bit for upper layer
+    **/
```
```suggestion
   // | Upper bits contain
   // | OID_FMT_TYPE_BITS (object type)
   // | OID_FMT_META_BITS (object class metadata)
   // | OID_FMT_CLASS_BITS (object class)
   // | 96-bit for upper layer
```

glmiotto commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -201,33 +212,27 @@ enum {
+
+/** Flags for oclass hints */
+enum {
+   /** Flags to control OC Redundancy */
+   DAOS_OCH_RDD_DEF = (1 << 0), /** Default - use RF prop */
+   DAOS_OCH_RDD_NO = (1 << 1),  /** No redundancy */
+   DAOS_OCH_RDD_RP = (1 << 2),  /** Replication */
+   DAOS_OCH_RDD_EC = (1 << 3),  /** Erasure Code */
```
```suggestion
   /// Flags to control OC Redundancy
   DAOS_OCH_RDD_DEF = (1 << 0), // Default - use RF prop
   DAOS_OCH_RDD_NO = (1 << 1),  // No redundancy
   DAOS_OCH_RDD_RP = (1 << 2),  // Replication
   DAOS_OCH_RDD_EC = (1 << 3),  // Erasure Code
```

glmiotto commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -201,33 +212,27 @@ enum {
+#define DAOS_OCH_RDD_MASK (DAOS_OCH_RDD_MAX_VAL << DAOS_OCH_RDD_SHIFT)
+#define DAOS_OCH_SHD_MASK (DAOS_OCH_SHD_MAX_VAL << DAOS_OCH_SHD_SHIFT)
+
+/** Flags for oclass hints */
```
```suggestion
/// Flags for oclass hints
```

glmiotto commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -145,33 +147,42 @@ typedef struct {
+#define OID_FMT_CLASS_SHIFT (OID_FMT_TYPE_SHIFT - OID_FMT_CLASS_BITS)
+#define OID_FMT_META_SHIFT (OID_FMT_CLASS_SHIFT - OID_FMT_META_BITS)
+
+/** DAOS object type */
```
```suggestion
/// DAOS object type
```

jalopezg-r00t commented on tree/ntuple/v7/src/RDaos.cxx:
```diff
@@ -85,7 +82,10 @@ ROOT::Experimental::Detail::RDaosObject::RDaosObject(RDaosContainer &container,
-      daos_obj_generate_id(&oid, DAOS_OF_DKEY_UINT64 | DAOS_OF_AKEY_UINT64 /*| DAOS_OF_ARRAY_BYTE*/, cid.fCid, 0);
+      daos_obj_generate_oid(container.fContainerHandle, &oid,
+                            static_cast<daos_otype_t>(DAOS_OT_DKEY_UINT64 | DAOS_OT_AKEY_UINT64), cid.fCid,
+                            DAOS_OCH_RDD_DEF, 0);
```
We may also explicitly mention a sharding OC hint flag (although as per current DAOS version, it behaves the same if left unspecified).
```suggestion
                            DAOS_OCH_RDD_DEF | DAOS_OCH_SHD_DEF, 0);
```

jalopezg-r00t commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -90,15 +91,16 @@ typedef struct daos_event {
+int daos_eq_query(daos_handle_t eqh, daos_eq_query_t query, unsigned int nevents, daos_event_t **events);
+
+int daos_event_test(daos_event_t *ev, int64_t timeout, bool *flag);
+int daos_event_parent_barrier(daos_event_t *ev);
```
I guess these new declarations are required by the follow-up PR that includes changes in the use of EQs, right?

glmiotto replied:
Correct, these should have only been added in the next PR. Sorry about the spoilers!

jalopezg-r00t commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -201,33 +212,27 @@ enum {
+   DAOS_OCH_RDD_NO = (1 << 1),  // No redundancy
+   DAOS_OCH_RDD_RP = (1 << 2),  // Replication
+   DAOS_OCH_RDD_EC = (1 << 3),  // Erasure Code
+};
```
We may need to add `DAOS_OCH_SHD_DEF`, if eventually used in RDaos.cxx.  OTOH, some of the other enumerators can be removed (unused); the less lines we have to maintain here, the better.

jalopezg-r00t commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -238,13 +243,78 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags,
+ * \return	  true      Label meets length/format requirements
+ *			  false     Label is not valid length or format
+ */
+static inline bool daos_label_is_valid(const char *label)
```
I would vote for keeping mocks simple.  Therefore, I suggest to get rid of this function; mocks do not really distinguish UUIDs from labels (any string provided by the user is just used as a key in a map).

jalopezg-r00t commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -238,13 +243,78 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags,
 
 //////////////////////////////////////////////////////////////////////////////// daos_prop.h
-
+#include <ctype.h>
```
Why is this needed?  Maybe move to the beginning of the file together with other `#include`s.

glmiotto replied:
This is due to `isalnum` and `isxdigit` in `daos_label_is_valid` from `daos_prop.h`, which might be removed altogether (or simplified if used for development)

jalopezg-r00t commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -257,15 +327,13 @@ typedef struct {
-		 daos_event_t *ev);
-int daos_cont_open(daos_handle_t poh, const uuid_t uuid, unsigned int flags,
-	       daos_handle_t *coh, daos_cont_info_t *info, daos_event_t *ev);
+int daos_cont_create(daos_handle_t poh, uuid_t uuid, daos_prop_t *cont_prop, daos_event_t *ev);
```
We might even remove this decl (unused now); however, if we want to keep it, see the comment in `libdaos_mock.cxx`

glmiotto replied:
It would make sense to keep this declaration for development purposes if we do support UUID URIs in `RDaos` as well as label ones, even if the mock library and test suites don't use it.

jalopezg-r00t commented on tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:
```diff
@@ -299,91 +276,101 @@ int daos_oclass_id2name(daos_oclass_id_t oc_id, char *name)
-   std::copy_n(uuid, std::tuple_size<Uuid_t>::value, std::begin(u));
-   pool->CreateContainer(u);
+
+   pool->CreateContainer((label_t) reinterpret_cast<char *>(uuid));
```
Construct a temporary `label_t` (aka `std::string`) instead.  Same for other similar uses below.
```suggestion
   pool->CreateContainer(label_t(reinterpret_cast<char *>(uuid)));
```

jalopezg-r00t commented on tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:
```diff
@@ -299,91 +276,101 @@ int daos_oclass_id2name(daos_oclass_id_t oc_id, char *name)
+   if (!pool)
+      return -DER_INVAL;
+
+   if (!daos_label_is_valid(label))
+      return -DER_INVAL;
```
See comment in `libdaos_mock/daos.h`.

jalopezg-r00t commented on tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:
```diff
@@ -169,18 +158,18 @@ class RDaosFakeContainer {
    /// \brief Get a pointer to a RDaosFakePool object associated to the given UUID.
    /// Non-existent pools shall be created on-demand.
-   static RDaosFakePool *GetPool(const Uuid_t uuid)
+   static RDaosFakePool *GetPool(const label_t label)
```
We should probably take a reference type instead (same for other uses, e.g. in `CreateContainer()`).
```suggestion
   static RDaosFakePool *GetPool(const label_t &label)
```

jalopezg-r00t reviewed (changes requested):
Thanks!  Very nice work, @glmiotto :slightly_smiling_face:!

I have added some comments that we have to address before merging.

glmiotto commented:
> Thanks! Very nice work, @glmiotto slightly_smiling_face!
> 
> I have added some comments that we have to address before merging.

> Thanks! Very nice work, @glmiotto :slightly_smiling_face:!
> 
> I have added some comments that we have to address before merging.

Great feedback @jalopezg-r00t, thanks as always! I'll get right on it

glmiotto commented on tree/ntuple/v7/test/ntuple_storage_daos.cxx:
```diff
@@ -1,15 +1,16 @@
    diags.requiredDiag(kWarning, "[ROOT.NTuple]", "Pre-release format version: RC 1", false);
 
-   std::string daosUri("daos://" R__DAOS_TEST_POOL ":1/a947484e-e3bc-48cb-8f71-3292c19b59a4");
+   std::string daosUri("daos://" R__DAOS_TEST_POOL "/container-test-1");
```
Well thought, will do

glmiotto commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -201,33 +212,27 @@ enum {
+
+/// Flags for oclass hints
+enum {
+   /// Flags to control OC Redundancy
+   DAOS_OCH_RDD_DEF = (1 << 0), // Default - use RF prop
+   DAOS_OCH_RDD_NO = (1 << 1),  // No redundancy
+   DAOS_OCH_RDD_RP = (1 << 2),  // Replication
+   DAOS_OCH_RDD_EC = (1 << 3),  // Erasure Code
```
```suggestion
   // OC Redundancy flags
   DAOS_OCH_RDD_DEF = (1 << 0), // Default: RF prop
   // OC Sharding flags
   DAOS_OCH_SHD_DEF = (1 << 4), // Default: MAX for array & flat KV, else 1
```

glmiotto commented on tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:
```diff
@@ -299,91 +276,101 @@ int daos_oclass_id2name(daos_oclass_id_t oc_id, char *name)
+   if (!daos_label_is_valid(label))
+      return -DER_INVAL;
+
+   pool->CreateContainer((label_t)label);
```
```suggestion
   pool->CreateContainer(label_t(label));
```

glmiotto commented on tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:
```diff
@@ -299,91 +276,101 @@ int daos_oclass_id2name(daos_oclass_id_t oc_id, char *name)
+   if (!daos_label_is_valid(label))
+      return -DER_INVAL;
+
+   auto cont = pool->GetContainer((label_t)label);
```
```suggestion
   auto cont = pool->GetContainer(label_t(label));
```

glmiotto commented on tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:
```diff
@@ -428,24 +428,16 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags, daos_key
-   std::copy_n(uuid, std::tuple_size<Uuid_t>::value, std::begin(u));
-   *poh = RDaosHandle::ToHandle(RDaosFakePool::GetPool(u));
+
+   *poh = RDaosHandle::ToHandle(RDaosFakePool::GetPool((label_t)label));
```
```suggestion
   *poh = RDaosHandle::ToHandle(RDaosFakePool::GetPool(label_t(label)));
```

glmiotto commented on tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:
```diff
@@ -189,22 +178,22 @@ class RDaosFakePool {
    ~RDaosFakePool() = default;
 
-   void CreateContainer(const Uuid_t uuid)
+   void CreateContainer(const label_t label)
```
```suggestion
   void CreateContainer(const label_t &label)
```

glmiotto commented on tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:
```diff
@@ -189,22 +178,22 @@ class RDaosFakePool {
    }
 
-   RDaosFakeContainer *GetContainer(const Uuid_t uuid)
+   RDaosFakeContainer *GetContainer(const label_t label)
```
```suggestion
   RDaosFakeContainer *GetContainer(const label_t &label)
```

glmiotto commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -90,15 +91,16 @@ typedef struct daos_event {
-int daos_eq_poll(daos_handle_t eqh, int wait_running,
-	     int64_t timeout, unsigned int nevents, daos_event_t **events);
+int daos_eq_poll(daos_handle_t eqh, int wait_running, int64_t timeout, unsigned int nevents, daos_event_t **events);
+int daos_eq_query(daos_handle_t eqh, daos_eq_query_t query, unsigned int nevents, daos_event_t **events);
+
+int daos_event_test(daos_event_t *ev, int64_t timeout, bool *flag);
+int daos_event_parent_barrier(daos_event_t *ev);
```
```suggestion
```

glmiotto commented:
@phsft-bot build just on ROOT-ubuntu16/nortcxxmod

glmiotto commented:
@phsft-bot build

jalopezg-r00t commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -239,12 +232,45 @@ int daos_obj_update(daos_handle_t oh, daos_handle_t th, uint64_t flags,
+ */
+static inline bool daos_label_is_valid(const char *label)
+{
+   int len;
+   int i;
+
+   if (label == NULL)
+      return false;
+
+   len = strnlen(label, DAOS_PROP_LABEL_MAX_LEN + 1);
+   if (len == 0 || len > DAOS_PROP_LABEL_MAX_LEN)
+      return false;
+
+   // Verify that it contains only alphanumeric characters or :.-_
+   for (i = 0; i < len; i++) {
+      char c = label[i];
+      if (isalnum(c) || c == '.' || c == '_' || c == ':' || c == '-')
+         continue;
+      return false;
+   }
+   return true;
```
In principle, and to keep the header small, we can even simply `return true;` here, as it is not important for the unit tests.

jalopezg-r00t commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -254,37 +280,35 @@ typedef struct {
 typedef struct {
-	char		unused; // silence [-Wextern-c-compat]
+   uuid_t ci_uuid;
+   char unused; // silence [-Wextern-c-compat]
```
This member was only added to silence the `-Wextern-c-compat`, where 0-sized objects are not required to have different addresses.  Thus, we can remove it now. 
```suggestion
```

jalopezg-r00t commented on tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:
```diff
@@ -254,37 +280,35 @@ typedef struct {
 typedef struct {
-	char		unused; // silence [-Wextern-c-compat]
+   uuid_t pi_uuid;
+   char unused; // silence [-Wextern-c-compat]
```
Same here.

jalopezg-r00t reviewed (changes requested):
Getting closer!  :slightly_smiling_face:

I think this will be the last round of feedback.  After that, we will probably need to squash some commits.

jalopezg-r00t reviewed (approved):
LGTM!
